GW2EIEvtcParser/EncounterLogic/Raids/W4/Deimos.cs

```diff
@@ -278,59 +278,83 @@ internal override void EIEvtcParse(FightData fightData, AgentData agentData, Lis
 
         internal override List<PhaseData> GetPhases(ParsedEvtcLog log, bool requirePhases)
         {
-            long start = 0;
-            long end = 0;
-            long fightDuration = log.FightData.FightEnd;
             List<PhaseData> phases = GetInitialPhase(log);
             NPC mainTarget = Targets.FirstOrDefault(x => x.ID == (int)ArcDPSEnums.TargetID.Deimos);
             if (mainTarget == null)
             {
                 throw new MissingKeyActorsException("Deimos not found");
             }
             phases[0].AddTarget(mainTarget);
-            if (!requirePhases)
+
+            if (requirePhases)
             {
-                return phases;
+                phases = AddBossPhases(phases, log, mainTarget);
+                phases = AddAddPhases(phases, log, mainTarget);
+                phases = AddBurstPhases(phases, log, mainTarget);
             }
+
+            return phases;
+        }
+
+        private List<PhaseData> AddBossPhases(List<PhaseData> phases, ParsedEvtcLog log, NPC mainTarget)
+        {
             // Determined + additional data on inst change
             AbstractBuffEvent invulDei = log.CombatData.GetBuffData(762).FirstOrDefault(x => x is BuffApplyEvent && x.To == mainTarget.AgentItem);
+
             if (invulDei != null)
             {
-                end = invulDei.Time;
-                phases.Add(new PhaseData(start, end));
-                start = _deimos10PercentTime > 0 ? _deimos10PercentTime : fightDuration;
+                var phase100to10 = new PhaseData(0, invulDei.Time, "100% - 10%");
+                phase100to10.AddTarget(mainTarget);
+                phases.Add(phase100to10);
+
+                if (_deimos10PercentTime > 0)
```

Comment:
If phases are not parsed, the health is never corrected.

Also the reasons why the hp modification was done in check CM were:
- Modifying the content of the agent/combat event outside of the EvtcParsedLog contructor is very bad practice, it is not necessary to call getPhases(...) so that would mean the log is in a bad state (event without counting the requirePhases conditioning)
- The modification only depends on if deimos10% was reached and isCM, both known after EIEvtcParse and IsCM respectively and both are called before EvtcParsedLog is created

I like the overall refactoring, that is always appreciated, but I very much prefer keeping "no agent or combat event modification done" once ParsedEvtcLog has been created as EIParser can be plugged into any other application for parsing purposes and I'm not in favor of possibly having non complete logs unless some public APIs have been called post constructor.

Reply:
You're right about health not being corrected if phases are not being parsed, I overlooked that use-case because I never turn phases off.

Regarding the comments about bad state, I can't fully comprehend the architectural implications behind it yet but I'll just believe you that it's bad.
I guess I was going too much along what I had seen for the GetPhases on Xera for correcting the health.
Wouldn't setting the health for Xera's 2nd phase cause a similar problem?

I looked up where the `IsCM` method gets called and found that it's called in the `SetCM` method of FightData.
I think that splitting the content of that method into 2 methods `IsCm` (only fetching the state) and a new interface method `ApplyCmChanges` (actually changing the health value) would make it more intuitive but given that Deimos seems to be the only use case for this, it's probably not a significant addition. 
For now I'll move it back into IsCM and add a comment with the reasoning why it's there.

Thanks for the feedback, I'll look to get it done in the next half hour.

Reply:
To give you an example, if we do it in GetPhases(). Your user will parse their log, then, for example, look at the targets list in order to find Deimos. They will then call GetHealth() on it in order to check its health for one reason or another. As that user has not called GetPhases() before hand, they'll get the non corrected health value and unless they call GetHealth() a second time after calling GetPhases() (not sure they'll even call it), they'll be stuck with a bad value. 
If we do it inside the ParsedEvtcLog constructor, where all that CM and specific EI processing/redirection stuff happens, then your user will never have to worry about the order in which they need to call public APIs to make sure the values they get are correct.

You are absolutely correct with your 2nd point. The code is pretty big, I do a lot of cleanup and refactoring but sometimes stuff gets past the net. I'll do a pass on name/hp override related stuff to make sure they are called in correct spots.

Regarding the 3rd, yes, that could be done. I did not particularly bother with it because it is very specific to Deimos, for the others it can always be done in EIEvtcParse where EncounterLogic specific operations are done. That would definitively be the path to take should additional encounters with such situations arise or if there is a need to do much more complex operations around CM conditions.

Reply:
Thank you for the explanation :-)